src/test/scala/sigmastate/SoftForkabilitySpecification.scala

```diff
@@ -0,0 +1,292 @@
+    val ctx = createContext(blockHeight, tx, vs)
+    val prop = tx.outputs(0).ergoTree
+    val proof1 = prover.prove(env, prop, ctx, fakeMessage).get
+    proof1
+  }
+
+  def verifyTx(name: String, tx: ErgoLikeTransaction, proof: ProverResult, vs: SigmaValidationSettings) = {
+    val env = Map(ScriptNameProp -> (name + "_verify"))
+    val ctx = createContext(blockHeight, tx, vs)
+    val prop = tx.outputs(0).ergoTree
+    verifier.verify(env, prop, ctx, proof, fakeMessage).map(_._1).fold(t => throw t, identity) shouldBe true
+  }
+
+  def proveAndVerifyTx(name: String, tx: ErgoLikeTransaction, vs: SigmaValidationSettings) = {
+    val proof = proveTx(name, tx, vs)
+    verifyTx(name, tx, proof, vs)
+  }
+
+  def checkTxProp[T <: ErgoLikeTransaction, R](tx1: T, tx2: T)(p: T => R) = {
+    p(tx1) shouldBe p(tx2)
+  }
+
+  property("node v1, received tx with script v1, incorrect script") {
+    assertExceptionThrown({
+      // CheckDeserializedScriptIsSigmaProp rule violated
+      ErgoLikeTransaction.serializer.parse(SigmaSerializer.startReader(invalidTxV1bytes))
+    }, {
+      case se: SerializerException if se.cause.isDefined =>
+        val ve = se.cause.get.asInstanceOf[ValidationException]
+        ve.rule == CheckDeserializedScriptIsSigmaProp
+      case _ => false
+    })
+  }
+
+  property("node v1, received tx with script v1, correct script") {
+    // able to parse
+    val tx = ErgoLikeTransaction.serializer.parse(SigmaSerializer.startReader(txV1bytes))
+
+    // validating script
+    proveAndVerifyTx("propV1", tx, vs)
+  }
+
+  val Height2Code = (LastConstantCode + 56).toByte
+  /** Same as Height, but new opcode to test soft-fork */
+  case object Height2 extends NotReadyValueInt {
+    override val opCode: OpCode = Height2Code // use reserved code
+    def opType = SFunc(SContext, SInt)
+  }
+  val Height2Ser = CaseObjectSerialization(Height2Code, Height2)
+
+  // prepare soft-fork settings for v2
+  val v2vs = vs.updated(CheckValidOpCode.id, ChangedRule(Array(Height2Code)))
+
+  /** Runs the block as if on the v2 node. */
+  def runOnV2Node[T](block: => T): T = {
+    ValueSerializer.addSerializer(Height2Code, Height2Ser)
+    val res = block
+    ValueSerializer.removeSerializer(Height2Code)
+    res
+  }
+
+  lazy val prop = GT(Height2, IntConstant(deadline))
+  lazy val invalidTxV2 = createTransaction(createBox(boxAmt, prop.asSigmaProp, 1))
+
+
+  lazy val propV2 = prop.toSigmaProp
+  // prepare bytes using special serialization WITH `size flag` in the header
+  lazy val propV2tree = ErgoTree.withSegregation(ErgoTree.SizeFlag,  propV2)
+  lazy val propV2treeBytes = runOnV2Node {
+    propV2tree.bytes
+  }
+
+  lazy val txV2 = createTransaction(createBox(boxAmt, propV2tree, 1))
+  lazy val txV2messageToSign = runOnV2Node {
+    txV2.messageToSign
+  }
+
+  val newTypeCode = (SGlobal.typeCode + 1).toByte
+
+  property("node v1, soft-fork up to v2, script v2 without size") {
```

Comment:
I propose to reorganize tests to ensure, that node v1 and node v2 have the same behaviour under some conditions.
In this particular case both nodes v1 and v2 should fail.